File: src/main/java/de/pflugradts/passbird/domain/model/password/PasswordEntry.java
package de.pflugradts.passbird.domain.model.password;

import de.pflugradts.passbird.domain.model.PasswordEntryCreated;
import de.pflugradts.passbird.domain.model.PasswordEntryDiscarded;
import de.pflugradts.passbird.domain.model.PasswordEntryRenamed;
import de.pflugradts.passbird.domain.model.PasswordEntryUpdated;
import de.pflugradts.passbird.domain.model.ddd.AggregateRoot;
import de.pflugradts.passbird.domain.model.ddd.DomainEvent;
import de.pflugradts.passbird.domain.model.namespace.NamespaceSlot;
import de.pflugradts.passbird.domain.model.transfer.Bytes;
import de.pflugradts.passbird.domain.service.password.storage.PasswordStoreAdapterPort;
import lombok.AccessLevel;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.ToString;

import java.util.ArrayList;
import java.util.List;

/**
 * A PasswordEntry represents a {@link Key} and an associated {@link Password} stored in the
 * {@link PasswordStoreAdapterPort PasswordStore}.
 */
@EqualsAndHashCode(of = {"key", "namespace"})
@ToString(of = {"key", "namespace"})
public class PasswordEntry implements AggregateRoot {

    @Getter(AccessLevel.PRIVATE)
    private final Key key;
    @Getter(AccessLevel.PRIVATE)
    private final Password password;
    @Getter(AccessLevel.PRIVATE)
    private NamespaceSlot namespace;
    @Getter
    private final List<DomainEvent> domainEvents = new ArrayList<>();

    protected PasswordEntry(final NamespaceSlot namespaceSlot, final Bytes keyBytes, final Bytes passwordBytes) {
        namespace = namespaceSlot;
        key = Key.Companion.createKey(keyBytes.copy());
        password = Password.Companion.createPassword(passwordBytes.copy());
        registerDomainEvent(new PasswordEntryCreated(this));
    }

    public static PasswordEntry create(final NamespaceSlot namespaceSlot,
                                       final Bytes keyBytes,
                                       final Bytes passwordBytes) {
        return new PasswordEntry(namespaceSlot, keyBytes, passwordBytes);
    }

    public NamespaceSlot associatedNamespace() {
        return namespace;
    }

    public Bytes viewKey() {
        return getKey().view();
    }

    public Bytes viewPassword() {
        return getPassword().view();
    }

    public void rename(final Bytes keyBytes) {
        getKey().rename(keyBytes);
        registerDomainEvent(new PasswordEntryRenamed(this));
    }

    public void updatePassword(final Bytes bytes) {
        getPassword().update(bytes);
        registerDomainEvent(new PasswordEntryUpdated(this));
    }

    public void updateNamespace(final NamespaceSlot namespaceSlot) {
        this.namespace = namespaceSlot;
    }

    public void discard() {
        getPassword().discard();
        registerDomainEvent(new PasswordEntryDiscarded(this));
    }

}

File: src/main/kotlin/de/pflugradts/passbird/domain/model/password/PasswordEntry.kt
package de.pflugradts.passbird.domain.model.password

import de.pflugradts.passbird.domain.model.ddd.AggregateRoot
import de.pflugradts.passbird.domain.model.event.PasswordEntryCreated
import de.pflugradts.passbird.domain.model.event.PasswordEntryDiscarded
import de.pflugradts.passbird.domain.model.event.PasswordEntryRenamed
import de.pflugradts.passbird.domain.model.event.PasswordEntryUpdated
import de.pflugradts.passbird.domain.model.nest.Slot
import de.pflugradts.passbird.domain.model.password.Key.Companion.createKey
import de.pflugradts.passbird.domain.model.password.Password.Companion.createPassword
import de.pflugradts.passbird.domain.model.transfer.Bytes

/**
 * A PasswordEntry represents a [Key] and an associated [Password] stored in the
 * [PasswordStore][PasswordStoreAdapterPort].
 */
class PasswordEntry private constructor(
    private var nestSlot: Slot,
    private val key: Key,
    private val password: Password,
) : AggregateRoot() {

    init { registerDomainEvent(PasswordEntryCreated(this)) }
    fun associatedNest() = nestSlot
    fun viewKey() = key.view()
    fun viewPassword() = password.view()

    fun rename(keyBytes: Bytes) {
        key.rename(keyBytes)
        registerDomainEvent(PasswordEntryRenamed(this))
    }

    fun updatePassword(bytes: Bytes) {
        password.update(bytes)
        registerDomainEvent(PasswordEntryUpdated(this))
    }

    fun moveToNestAt(slot: Slot) { nestSlot = slot }

    fun discard() {
        password.discard()
        registerDomainEvent(PasswordEntryDiscarded(this))
    }

    override fun equals(other: Any?) =
        (other as? PasswordEntry)?.let { it.viewKey() == viewKey() && it.associatedNest() == nestSlot } ?: false
    override fun hashCode() = nestSlot.hashCode() + 31 * key.hashCode()

    companion object {
        fun createPasswordEntry(slot: Slot, keyBytes: Bytes, passwordBytes: Bytes) =
            PasswordEntry(slot, createKey(keyBytes), createPassword(passwordBytes))
    }
}
